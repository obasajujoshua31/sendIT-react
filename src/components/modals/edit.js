import React, { Component } from "react";
import ReactModal from "react-modal";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";
import getOneParcel from "../../actions/parcels/getOneParcel";
import Spinner from "react-spinkit";
import EditDestinationModal from "./editDestination";
import CancelParcelOrder from "./cancelOrder";
import ChangeLocation from "./location";
import UpdateStatus from "./status";

ReactModal.setAppElement("#app");

class Modal extends Component {
  constructor(props) {
    super(props);
    this.state = {
      parcel: [],
      showEditModal: false,
      showCancelModal: false
    };
    this.editParcel = this.editParcel.bind(this);
    this.closeParcelModal = this.closeParcelModal.bind(this);
    this.showCancelOrder = this.showCancelOrder.bind(this);
    this.closeCancelModal = this.closeCancelModal.bind(this);
  }

  editParcel() {
    this.setState({ showEditModal: true });
  }
  closeParcelModal() {
    this.setState({ showEditModal: false });
  }
  showCancelOrder() {
    this.setState({ showCancelModal: true });
  }
  closeCancelModal() {
    this.setState({ showCancelModal: false });
  }
  componentDidMount() {
    this.props.loadOneParcel(this.props.parcelId, this.props.role);
  }
  componentDidUpdate(prevProps) {
    if (prevProps !== this.props && this.props.userParcel.data) {
      this.setState({ parcel: this.props.userParcel.data });
    }
  }
  render() {
    const { show, handleCloseModal } = this.props;

    if (!this.state.parcel.length) {
      return <Spinner name="three-bounce" color="orange" />;
    }

    const {
      parcel_name,
      destination,
      pick_up_location,
      sent_on,
      delivered_on,
      weight_metric,
      weight,
      status,
      present_location,
      parcel_id
    } = this.state.parcel[0];

    const modal = this.props.role ? (
      <ChangeLocation
        show={this.state.showEditModal}
        closeChangeLocation={this.closeParcelModal}
        parcelId={parcel_id}
        presentLocation={present_location}
      />
    ) : (
      <EditDestinationModal
        show={this.state.showEditModal}
        closeEdit={this.closeParcelModal}
        parcelId={parcel_id}
        destination={destination}
      />
    );

    const cancelModal = this.props.role ? (
      <UpdateStatus
        show={this.state.showCancelModal}
        closeCancel={this.closeCancelModal}
        parcelId={parcel_id}
        status={status}
      />
    ) : (
      <CancelParcelOrder
        show={this.state.showCancelModal}
        closeCancel={this.closeCancelModal}
        parcelId={parcel_id}
        status={status}
      />
    );

    const isCancellable = status === "PLACED" || status === "TRANSITING";
    return (
      <div>
        <ReactModal
          isOpen={show}
          contentLabel="Modal #1 Global Style Override Example"
          onRequestClose={handleCloseModal}
          overlayClassName="Overlay"
          className="Modal"
        >
          <div style={modalHeaderStyle}>
            <span style={spanStyle}>
              <i className="far fa-times-circle" onClick={handleCloseModal} />
            </span>
            <h3 style={modalHeadBrandStyle}>Order</h3>
          </div>
          <div style={modalBodyContainer}>
            <div style={modalBodyStyle}>
              <div style={itemStyle}>
                <label className="item-label">Name:</label>
                <span className="item-span">{parcel_name}</span>
              </div>
              <hr />
              <div style={itemStyle}>
                <label className="item-label">Pick Up Location:</label>
                <span className="item-span">{pick_up_location}</span>
              </div>
              <hr />
              <div style={itemStyle}>
                <label className="item-label">Destination:</label>
                <span className="item-span">{destination}</span>
              </div>
              <hr />
              <div style={itemStyle}>
                <label className="item-label">Status: </label>
                <span className="item-span">{status}</span>
              </div>
              <hr />
              <div style={itemStyle}>
                <label className="item-label">Sent On:</label>
                <span className="item-span">{sent_on}</span>
              </div>
              <hr />

              <div style={itemStyle}>
                <label className="item-label">Weight:</label>
                <span className="item-span">{`${weight}  ${weight_metric}`}</span>
              </div>
              <hr />
              <div style={itemStyle}>
                <label className="item-label">Delivered On:</label>
                <span className="item-span">{delivered_on}</span>
              </div>
              <hr />
              <div style={itemStyle}>
                <label className="item-label">Present Location:</label>
                <span className="item-span">{present_location}</span>
              </div>
            </div>
          </div>

          <div style={modalFooterStyle}>
            <button
              className="button edit"
              onClick={this.editParcel}
              hidden={!isCancellable}
            >
              {this.props.role ? "Change Location" : "Edit Parcel"}
            </button>
            <button
              className="button cancel"
              onClick={this.showCancelOrder}
              hidden={!isCancellable}
            >
              {this.props.role ? "Change Status" : "Cancel Parcel"}
            </button>
          </div>
        </ReactModal>
        {this.state.showEditModal && modal}
        {this.state.showCancelModal && cancelModal}
      </div>
    );
  }
}

const spanStyle = {
  fontSize: "1.8em",
  color: "white",
  cursor: "pointer",
  padding: 5,
  position: "absolute",
  right: 5,
  bottom: 5
};

const modalHeaderStyle = {
  padding: 15,
  backgroundColor: "#806B75",
  position: "relative",
  display: "flex",
  justifyContent: "center",
  alignItem: "center"
};
const modalHeadBrandStyle = {
  fontWeight: "bold",
  fontSize: "1.4rem",
  color: "white"
};
const modalBodyStyle = {
  width: "100%",
  padding: 10,
  border: "2px solid rgba(0, 0, 0, 0.3)",
  borderRadius: 10,
  marginTop: 30,
  overflow: "scroll",
  height: "300px"
};
const itemStyle = {
  display: "flex",
  padding: 5
};
const modalFooterStyle = {
  backgroundColor: "#e9e6e6",
  width: "100%",
  // position: "absolute",
  // bottom: "0",
  // height: "40px",
  display: "flex",
  padding: "10px 0px",
  justifyContent: "space-around"
};
const modalBodyContainer = {
  padding: "10px",
  maxWidth: "100%",
  display: "flex",
  justifyContent: "center",
  alignItem: "center"
};

const mapStateToProps = ({ parcels }) => ({
  userParcel: parcels.parcel
});

const mapDispatchToProps = dispatch => ({
  loadOneParcel: (parcelId, role) => dispatch(getOneParcel(parcelId, role))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withRouter(Modal));
