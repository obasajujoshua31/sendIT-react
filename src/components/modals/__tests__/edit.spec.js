import React from "react";
import { shallow } from "enzyme";
import sinon from "sinon";
import { Modal, mapDispatchToProps, mapStateToProps } from "../edit";
import { wrap } from "module";

const props = {
  parcelId: "1",
  role: false,
  userParcel: {
    data: []
  },
  loadOneParcel: jest.fn(),
  show: true,
  handleCloseModal: jest.fn()
};
const mockStore = {
  parcels: {
    parcel: {
      data: ["fish"]
    }
  }
};

const mockState = {
  parcel: [
    {
      parcel_name: "phone",
      destination: "dkkdkd",
      pick_up_location: "dkld",
      sent_on: "slldd",
      delivered_on: "ldlldd",
      weight_metric: "kg",
      weight: 32,
      status: "PLACED",
      present_location: "dlddld",
      parcel_id: 1
    }
  ]
};

const wrapper = shallow(<Modal {...props} />);

describe("Edit Modal Test", () => {
  it("should render Edit Modal", () => {
    expect(wrapper).toMatchSnapshot();
  });
  it("dispatch should call loadOneParcel", () => {
    const dispatchSpy = sinon.spy();
    const dispatchResult = mapDispatchToProps(dispatchSpy);
    dispatchResult.loadOneParcel("1", false);
    expect(dispatchSpy.callCount).toEqual(1);
  });

  it("map state to props should return parcels", () => {
    const mapStateResult = mapStateToProps(mockStore);
    expect(mapStateResult.userParcel).toEqual({ data: ["fish"] });
  });

  it("editParcel should set State to true", () => {
    const editParcelSpy = jest.spyOn(wrapper.instance(), "editParcel");
    editParcelSpy();
    expect(wrapper.instance().state.showEditModal).toEqual(true);
  });

  it("closeParcelModal should set State to false", () => {
    const editParcelSpy = jest.spyOn(wrapper.instance(), "closeParcelModal");
    editParcelSpy();
    expect(wrapper.instance().state.showEditModal).toEqual(false);
  });

  it("showCancelOrder should set State to true", () => {
    const editParcelSpy = jest.spyOn(wrapper.instance(), "showCancelOrder");
    editParcelSpy();
    expect(wrapper.instance().state.showCancelModal).toEqual(true);
  });

  it("closeCancelModal should set State to false", () => {
    const editParcelSpy = jest.spyOn(wrapper.instance(), "closeCancelModal");
    editParcelSpy();
    expect(wrapper.instance().state.showCancelModal).toEqual(false);
  });

  it("should render when the role is false", () => {
    wrapper.setProps({
      role: false
    });
    expect(wrapper).toMatchSnapshot();
  });

  it("should render when the role is true", () => {
    wrapper.setProps({
      role: true
    });
    expect(wrapper).toMatchSnapshot();
  });

  it("should render modal when parcel is in the state", () => {
    wrapper.setState({
      parcel: mockState.parcel
    });
    expect(wrapper).toMatchSnapshot();
  });
});
