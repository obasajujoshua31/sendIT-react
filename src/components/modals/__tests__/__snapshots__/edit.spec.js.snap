// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Modal Test should render Edit Modal 1`] = `
<MDSpinner
  color="orange"
  color1="rgb(66, 165, 245)"
  color2="rgb(239, 83, 80)"
  color3="rgb(253, 216, 53)"
  color4="rgb(76, 175, 80)"
  duration={1333}
  name="three-bounce"
  size={28}
/>
`;

exports[`Edit Modal Test should render modal when parcel is in the state 1`] = `
<div>
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    className="Modal"
    closeTimeoutMS={0}
    contentLabel="Modal #1 Global Style Override Example"
    isOpen={true}
    onRequestClose={[MockFunction]}
    overlayClassName="Overlay"
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    role="dialog"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
  >
    <div
      style={
        Object {
          "alignItem": "center",
          "backgroundColor": "#806B75",
          "display": "flex",
          "justifyContent": "center",
          "padding": 15,
          "position": "relative",
        }
      }
    >
      <span
        style={
          Object {
            "bottom": 5,
            "color": "white",
            "cursor": "pointer",
            "fontSize": "1.8em",
            "padding": 5,
            "position": "absolute",
            "right": 5,
          }
        }
      >
        <i
          className="far fa-times-circle"
          onClick={[MockFunction]}
        />
      </span>
      <h3
        style={
          Object {
            "color": "white",
            "fontSize": "1.4rem",
            "fontWeight": "bold",
          }
        }
      >
        Order
      </h3>
    </div>
    <div
      style={
        Object {
          "alignItem": "center",
          "display": "flex",
          "justifyContent": "center",
          "maxWidth": "100%",
          "padding": "10px",
        }
      }
    >
      <div
        style={
          Object {
            "border": "2px solid rgba(0, 0, 0, 0.3)",
            "borderRadius": 10,
            "height": "300px",
            "marginTop": 30,
            "overflow": "scroll",
            "padding": 10,
            "width": "100%",
          }
        }
      >
        <div
          style={
            Object {
              "display": "flex",
              "padding": 5,
            }
          }
        >
          <label
            className="item-label"
          >
            Name:
          </label>
          <span
            className="item-span"
          >
            phone
          </span>
        </div>
        <hr />
        <div
          style={
            Object {
              "display": "flex",
              "padding": 5,
            }
          }
        >
          <label
            className="item-label"
          >
            Pick Up Location:
          </label>
          <span
            className="item-span"
          >
            dkld
          </span>
        </div>
        <hr />
        <div
          style={
            Object {
              "display": "flex",
              "padding": 5,
            }
          }
        >
          <label
            className="item-label"
          >
            Destination:
          </label>
          <span
            className="item-span"
          >
            dkkdkd
          </span>
        </div>
        <hr />
        <div
          style={
            Object {
              "display": "flex",
              "padding": 5,
            }
          }
        >
          <label
            className="item-label"
          >
            Status: 
          </label>
          <span
            className="item-span"
          >
            PLACED
          </span>
        </div>
        <hr />
        <div
          style={
            Object {
              "display": "flex",
              "padding": 5,
            }
          }
        >
          <label
            className="item-label"
          >
            Sent On:
          </label>
          <span
            className="item-span"
          >
            slldd
          </span>
        </div>
        <hr />
        <div
          style={
            Object {
              "display": "flex",
              "padding": 5,
            }
          }
        >
          <label
            className="item-label"
          >
            Weight:
          </label>
          <span
            className="item-span"
          >
            32  kg
          </span>
        </div>
        <hr />
        <div
          style={
            Object {
              "display": "flex",
              "padding": 5,
            }
          }
        >
          <label
            className="item-label"
          >
            Delivered On:
          </label>
          <span
            className="item-span"
          >
            ldlldd
          </span>
        </div>
        <hr />
        <div
          style={
            Object {
              "display": "flex",
              "padding": 5,
            }
          }
        >
          <label
            className="item-label"
          >
            Present Location:
          </label>
          <span
            className="item-span"
          >
            dlddld
          </span>
        </div>
      </div>
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#e9e6e6",
          "display": "flex",
          "justifyContent": "space-around",
          "padding": "10px 0px",
          "width": "100%",
        }
      }
    >
      <button
        className="button edit"
        hidden={false}
        onClick={
          [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
      >
        Change Location
      </button>
      <button
        className="button cancel"
        hidden={false}
        onClick={
          [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
      >
        Change Status
      </button>
    </div>
  </Modal>
</div>
`;

exports[`Edit Modal Test should render when the role is false 1`] = `
<MDSpinner
  color="orange"
  color1="rgb(66, 165, 245)"
  color2="rgb(239, 83, 80)"
  color3="rgb(253, 216, 53)"
  color4="rgb(76, 175, 80)"
  duration={1333}
  name="three-bounce"
  size={28}
/>
`;

exports[`Edit Modal Test should render when the role is true 1`] = `
<MDSpinner
  color="orange"
  color1="rgb(66, 165, 245)"
  color2="rgb(239, 83, 80)"
  color3="rgb(253, 216, 53)"
  color4="rgb(76, 175, 80)"
  duration={1333}
  name="three-bounce"
  size={28}
/>
`;
